util.py

```diff
@@ -1025,3 +1027,46 @@ def matching_author_string(origName):
 
     return response
 
+
+def majority_ascii(s, threshold=0.5):
+    return sum([char.isascii() for char in s]) / len(s) > threshold
+
+
+def punctuation_density(words):
+    return sum([word[-1] in string.punctuation for word in words if word]) / len(words)
+
+
+def detect_language_from_abstract_and_title(
+    abstract_words, title, probability_threshold=0.7, punctuation_density_threshold=0.5
+):
+    # use some heuristics to catch some edge cases:
+    # - detect language from abstract if probability is high and abstract isn't too short (for abstracts with ascii/latin characters)
+    #   - (also check for punctuation density in abstract, to catch cases where the abstract is a punctuation-separated list of author names)
+    # - otherwise try title, with similar rules
+    # - give up and return None if neither of those work
+    DetectorFactory.seed = 0
+
+    try:
+        if abstract_words:
+            abstract = " ".join(abstract_words)
+            abstract_language = detect_langs(abstract)
+            if abstract_language and abstract_language[0]:
+                if (
+                    (len(abstract) > 20 or not majority_ascii(abstract))
```

Comment:
This tripped me up for a minute - I wanted it to be _len() > 20 *and* majority_ascii_. i.e., we want it to be long enough that being majority ascii is meaningful. But the _punctuation_density_ check does that job - majority ascii isn't the same as majority text. So if we want to analyze, e.g., kanji where len() > 20 isn't a useful check this looks like the way to do it.

Reply:
Yes, that was my thought; it seems to me that in a lot of alphabets with non-latin characters, there may be more information crammed into a smaller number of characters, so we shouldn't impose the length limit.